File: python/otp.py
import hmac
from base64 import b32decode, b32encode
from datetime import datetime, timezone
from hashlib import sha1
from os import urandom
from typing import Optional

KEY_SIZE = 64


def dt2unix(dt: datetime, divider: int = 30) -> int:
    """Convert dt datetime to unix timestamp and divide it by a given divider."""
    return int((dt - datetime(1970, 1, 1, tzinfo=timezone.utc)).total_seconds()) // divider


def secret(size: int = 0) -> str:
    """Generate a random secret key."""
    rnd_bytes = urandom(size or KEY_SIZE)
    return b32encode(rnd_bytes).decode()


def code(value: str, counter: Optional[int] = None) -> str:
    """Generate OTP code for a given value and counter."""
    if counter is None:
        counter = dt2unix(datetime.utcnow()) // 30

    key = b32decode(value.upper().encode())
    counter_bytes = counter.to_bytes(8, byteorder='big')

    m = hmac.new(key, digestmod=sha1)
    m.update(counter_bytes)
    digest = m.digest()

    offset = digest[-1] & 0xf
    truncated = int.from_bytes(digest[offset:offset + 4], byteorder='big') & 0x7fffffff

    # add left padding zeros and take right 6 digits
    return f'{truncated:06d}'[-6:]

File: go/otp.go
// Package otp contains the implementation of the One Time Password (OTP) algorithm.
// It's done by an article from https://zserge.com/posts/one-time-passwords/
package otp

import (
	"bytes"
	"crypto/hmac"
	crand "crypto/rand"
	"crypto/sha1"
	"encoding/base32"
	"encoding/binary"
	"fmt"
	"math/rand"
	"time"
)

// KeySize is the size of the secret key in bytes.
const KeySize = 64

// CryptoRandSource represents a source of uniformly-distributed random int64 values in the range [0, 1<<63).
type CryptoRandSource struct{}

// Int63 returns a non-negative random 63-bit integer as an int64 from CryptoRandSource.
func (CryptoRandSource) Int63() int64 {
	var b [8]byte
	_, err := crand.Read(b[:])
	if err != nil {
		panic(err) // fail - can't continue
	}
	return int64(binary.LittleEndian.Uint64(b[:]) & (1<<63 - 1))
}

// Seed is fake CryptoRandSource Seed implementation for Source interface.
func (CryptoRandSource) Seed(int64) {}

// Secret generates a new secret key.
func Secret(size int) (string, error) {
	var r rand.Source = CryptoRandSource{}
	if size == 0 {
		size = KeySize
	}

	b := make([]byte, size)

	_, err := rand.New(r).Read(b)
	if err != nil {
		return "", err
	}
	return base32.StdEncoding.EncodeToString(b), nil
}

// Code generates a new code for the given secret key.
func Code(secret string, counter int64) (string, error) {
	if counter < 0 {
		counter = time.Now().Unix() / 30
	}

	// decode Base-32 secret key
	key := make([]byte, KeySize)
	if _, err := base32.StdEncoding.Decode(key, bytes.ToUpper([]byte(secret))); err != nil {
		return "", err
	}

	// write counter as 8 bytes, big-endian
	b := make([]byte, 8)
	binary.BigEndian.PutUint64(b, uint64(counter))

	hash := hmac.New(sha1.New, key)
	hash.Write(b)
	b = hash.Sum(nil)

	// find offset, lower 4 bits of the last byte
	offset := b[len(b)-1] & 0xF
	// read 4 bytes from offset as 32-bit integer
	n := binary.BigEndian.Uint32(b[offset : offset+4])

	// covert it to decimal and return last 6 digits
	s := fmt.Sprintf("%06d", int(n&0x7FFFFFFF))
	return s[len(s)-6:], nil
}
